### PR TITLE
initial impl of in-mem sync client for testing.

### PR DESCRIPTION
**Pull request by raulk**
This PR introduces a `sync.Interface` interface that testground components can use, so that in tests we can inject an in-memory implementation that does not require Redis.

This PR includes an initial, untested (!) stab at an in-memory implementation.

It also places all composite operations (sugar operations) into a struct that we then embed in various clients.

## TODO

- [ ] unit tests.

### CODE REVIEW

File: sync/interface.go.go [hunk 1]
```diff
@@ -0,0 +1,24 @@
...
+	"context"
+	"io"
+)
+
+// Interface is used by Testground components to enable mocking/stubbing in
+// tests. Consequently, it doesn't have the Must* variants, as those are useful
+// for test plans only.
+//
+// If you're a test plan writer, you should not be using this.
+type Interface interface {
```
Comment (nonsense) [inline sync/interface.go.go]: `Interface` for a name sounds weird. Wouldn't it be better to call this `Service` ? Basically then we have multiple impl. of the `sync.Service` ?

File: sync/interface.go.go [hunk 2]
```diff
@@ -0,0 +1,24 @@
+package sync
```
Comment (nonsense) [inline sync/interface.go.go]: Filename `inteface.go.go` -> `interface.go` or `service.go`.

Comment (raulk) [reply sync/interface.go.go]: Please don't gooooo. IDE woes.

File: sync/client_inmem.go [hunk 3]
```diff
@@ -0,0 +1,126 @@
...
+	states        map[State]int
+	barriers      map[State][]*Barrier
+	subscriptions map[*Topic][]reflect.Value
+	published     map[*Topic][]interface{}
+}
+
+var _ Interface = (*inmemClient)(nil)
+
+// NewInmemClient creates an in-memory sync client for testing.
+func NewInmemClient() Interface {
```
Comment (nonsense) [inline sync/client_inmem.go]: I think you could just return the *inmemClient - `accept interfaces, return structs`, although this isn't a hard rule.

File: sync/interface.go.go (hunk 1 above)
Comment (raulk) [reply sync/interface.go.go]: But it's not really the _service_... I see why the name can be ugly, but it's idiomatic in go: https://golang.org/pkg/sort/#Interface :-/

File: sync/client_inmem.go (hunk 3 above)
Comment (raulk) [reply sync/client_inmem.go]: Yup. I wonder how it would show up in godocs though.

Comment (raulk) [reply sync/client_inmem.go]: image

... where obviously `*inmemClient` is not documented. I don't think that's nice.

File: sync/interface.go.go (hunk 1 above)
Comment (nonsense) [reply sync/interface.go.go]: It is the interface of the sync service - precisely what we a mocking for in-memory usage, no? In any case I don't have strong feelings on how we call this.

Comment (raulk) [reply sync/interface.go.go]: Yeah, but introducing the term `Service` to abstract over a `Client` is rather weird. Those names tend to have well-defined expectations.

Comment (nonsense) [reply sync/interface.go.go]: I agree, that this is really the `Client` of the `sync service`. At the moment we don't have a separate process for the `sync service` concept, it is really just a library interacting with Redis... This is why I think of the `Client` and the `Service` as the same thing right now.

In the future, if we split the `sync service` into a separate component, and it becomes not just a library that testplans import, but rather import a `Client` that interacts with the `Service` - then they would really be different things.

Comment (raulk) [reply sync/interface.go.go]: Yup, in that case it's very likely that the `Service` will not live in the SDK, even.

**Comment by Robmat05** [comment]
relates to https://github.com/testground/testground/pull/939

**Comment by coryschwartz** [comment]
I made a substantial number of edits in https://github.com/testground/sdk-go/pull/16/commits/fe2dafaa6dba74fe3355cee2971decb7b0e6d59e

In particular, the most substantive are these: 
*  I filled out the Interface to include Must-* methods. Maybe the tests in this package don't use them, but the sidecar does, and they are used under normal circumstances
* More code moved from the real *Client to *sugarOptions. shared "sugar" logic is there and the core implementation is specific to the real *Client or the InMem version.
* Changed the InMem client to to use map of strings, keyed by topic.name rather than map of *Topic. The previous method was simply broken, as it was keyed by pointer. If someone wants this to be keyed by Topic (the struct) this would probably work as well and I wouldn't disagree.
* Changed network.Client to use sync.Interface as for its sync client. This is again so the network client can be used with the InMem version.

File: sync/interface.go.go (hunk 1 above)
Comment (coryschwartz) [reply sync/interface.go.go]: I could see this being called `Syncer` or something. Most of the time when I see an interface simply called Interface, it is an interface to be implemented. i.e. sync.Interface, heap.Interface. Either way works for me though @nonsense if you have an opinion about it we can change this.

File: sync/barrier_test.go [hunk 4]
```diff
@@ -161,6 +161,29 @@ func TestBarrierDeadline(t *testing.T) {
 	}
 }
 
+func TestInmemSignalAndWait(t *testing.T) {
+	ctx, cancel := context.WithTimeout(context.Background(), 5*time.Second)
+	defer cancel()
+
+	runenv, cleanup := runtime.RandomTestRunEnv(t)
+	t.Cleanup(cleanup)
```
Comment (nonsense) [inline sync/barrier_test.go:169]: Nice one, I wasn't aware of that. We could refactor in a future PR the `RandomTestRunEnv` to not return a `cleanup` function at all, and jus use `t.Cleanup` inside.

**Review by nonsense — APPROVED** [review]

**Pull request merged**